### PR TITLE
ch3: misc fixes for undefined behavior sanitizer

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Fixes to clean up ch3 `ubsan` tests. Ref. https://jenkins-pmrs.cels.anl.gov/job/mpich-main-sanitizers/

The sock channel should be clean after this PR.

The nemesis channel still depends on PR #4481 to be clean.

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch3:sock
config: ubsan

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch3:tcp
config: ubsan

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch3:tcp
config: asan

**Comment by hzhou** [comment]
Re-test to confirm:
test:mpich/ch3/tcp

File: configure.ac [hunk 1]
```diff
@@ -2936,10 +2936,8 @@ else
     AC_DEFINE_UNQUOTED(MAX_ALIGNMENT,$ac_cv_alignof_long_double,[Controls byte alignment of structures (for aligning allocated structures)])
 fi
 
-# Require strict alignment memory access for alignment-sensitive platform (e.g., SPARC)
-if test "$host_cpu" = "sparc" ; then
-    AC_DEFINE(NEEDS_STRICT_ALIGNMENT,1,[Define if strict alignment memory access is required])
-fi
+# we need observe strict alignment to pass ubsan check
+AC_DEFINE(NEEDS_STRICT_ALIGNMENT,1,[Define if strict alignment memory access is required])
```
[the remark this answers is not in the document]
Comment (hzhou) [reply configure.ac:2940]: I guess we can add a configure options and use it for ubsan tests. The fixes went in ch4 is essentially already assuming this flag always true. I'll need to modify them to observe this flag as well.

[the remark this answers is not in the document]
Comment (raffenet) [reply configure.ac:2940]: Hmm, in that case it probably makes sense to make this the default. Perhaps we can override in the `--enable-fast` settings if on x86?

[the remark this answers is not in the document]
Comment (hzhou) [reply configure.ac:2940]: I think that sounds good. I'll add a commit to do that.

[the remark this answers is not in the document]
Comment (hzhou) [reply configure.ac]: Added last 2 commits.

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Pull request merged**